verticles/graphql/src/main/java/com/gentics/mesh/graphql/type/AbstractTypeProvider.java

```diff
@@ -563,20 +636,121 @@ protected DynamicStreamPageImpl<NodeContent> fetchFilteredNodes(DataFetchingEnvi
+				? nodeDao.findAllContent(project, gc, languageTags, type, pagingInfo, filters.getRight()) 
+				: nodeDao.findAllContent(project, gc, languageTags, type), 
+			pagingInfo, filters.getLeft(), (filters.getRight().isPresent() || PersistingRootDao.shouldSort(pagingInfo)) && PersistingRootDao.shouldPage(pagingInfo));
 	}
 
-	protected DynamicStreamPageImpl<NodeContent> applyNodeFilter(DataFetchingEnvironment env, Stream<? extends NodeContent> stream) {
+	public <T> Pair<Predicate<T>,Optional<FilterOperation<?>>> parseFilters(DataFetchingEnvironment env, EntityFilter<T> filterProvider) {
+		GraphQLContext gc = env.getContext();
 		Map<String, ?> filterArgument = env.getArgument("filter");
-		PagingParameters pagingInfo = getPagingInfo(env);
+		NativeQueryFiltering nativeQueryFiltering = options.getGraphQLOptions().getNativeQueryFiltering();
+		Predicate<T> javaFilter = null;
+		Optional<FilterOperation<?>> maybeNativeFilter = Optional.empty();
+
+		NativeFilter envNativeFilter = env.getArgumentOrDefault("nativeFilter", NativeFilter.IF_POSSIBLE);
+		if (filterArgument != null) {
+			switch (nativeQueryFiltering) {
+			case ON_DEMAND:
+				if (NativeFilter.NEVER.equals(envNativeFilter)) {
+					javaFilter = filterProvider.createPredicate(filterArgument);
+					break;
+				}
+				boolean invalid = true;
+				if (NativeFilter.ONLY.equals(envNativeFilter) || NativeFilter.IF_POSSIBLE.equals(envNativeFilter)) {
+					invalid = false;
+				}
+				if (invalid) {
+					throw new InvalidParameterException("Unsupported 'nativeFilter' parameter value: " + envNativeFilter);
+				}
+				// else fall through into the native filtering
+			case ALWAYS:
+				try {
+					maybeNativeFilter = Optional.of(filterProvider.createFilterOperation(filterArgument));
+					break;
+				} catch (UnformalizableQuery e) {
+					log.warn("The query filter cannot be formalized: {}", filterArgument);
+					log.debug(e);
+					if (NativeQueryFiltering.ALWAYS == nativeQueryFiltering || NativeFilter.ONLY.equals(envNativeFilter)) {
+						throw new InvalidParameterException(e.getLocalizedMessage());
+					} else {
+						log.warn("Trying to apply old Java filtering");
+					}// fall through into the old filtering
+				}			
+			case OFF:
+				if (NativeFilter.NEVER.equals(envNativeFilter)) {
+					throw new InvalidParameterException("Conflicting params: requested GraphQL 'nativeFilter' = " + envNativeFilter + ", Mesh GraphQL Options 'nativeQueryFiltering' = " + nativeQueryFiltering);
+				}
+				javaFilter = filterProvider.createPredicate(filterArgument);
+				break;
+			}
+		}
+		return Pair.of(javaFilter, maybeNativeFilter
+				.or(() -> Optional.of(PersistingRootDao.shouldPage(getPagingInfo(env)))
+						// We force native filtering, if paging is asked
+						.filter(b -> b && (NativeFilter.ONLY.equals(envNativeFilter) || NativeQueryFiltering.ALWAYS.equals(nativeQueryFiltering)))
+						.map(yes -> Comparison.dummy(true))));
```

Comment:
This does not compile (with the currently checked in version of graphql-filter)

Reply:
Argh. This is a '1 = 1' and '1 <> 1' operator.